src/contracts/core/DelegationManager.sol

```diff
@@ -224,50 +229,247 @@ contract DelegationManager is Initializable, OwnableUpgradeable, Pausable, Deleg
+
+        // Remove all strategies/shares from staker and operator and place into queue
+        return _removeSharesAndQueueWithdrawal({
+            staker: staker,
+            operator: operator,
+            withdrawer: staker,
+            strategies: strategies,
+            shares: shares
+        });
+    }
+
+    /**
+     * Allows a staker to withdraw some shares. Withdrawn shares/strategies are immediately removed
+     * from the staker. If the staker is delegated, withdrawn shares/strategies are also removed from
+     * their operator.
+     *
+     * All withdrawn shares/strategies are placed in a queue and can be fully withdrawn after a delay.
+     */
+     function queueWithdrawal(
+        IStrategy[] calldata strategies,
+        uint[] calldata shares,
+        address withdrawer
+    ) external onlyWhenNotPaused(PAUSED_ENTER_WITHDRAWAL_QUEUE) returns (bytes32) {
+        require(strategies.length == shares.length, "DelegationManager.queueWithdrawal: input length mismatch");
+        require(withdrawer != address(0), "DelegationManager.queueWithdrawal: must provide valid withdrawal address");
+        require(!isOperator(msg.sender), "DelegationManager.queueWithdrawal: operators cannot enter queue");
+
+        address operator = delegatedTo[msg.sender];
+
+        // Remove shares from staker's strategies and place strategies/shares in queue.
+        // If the staker is delegated to an operator, the operator's delegated shares are also reduced
+        // NOTE: This will fail if the staker doesn't have the shares implied by the input parameters
+        return _removeSharesAndQueueWithdrawal({
+            staker: msg.sender,
+            operator: operator,
+            withdrawer: withdrawer,
+            strategies: strategies,
+            shares: shares
+        });
+    }
+
+    function completeQueuedWithdrawal(
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint256 middlewareTimesIndex,
+        bool receiveAsTokens
+    ) external onlyWhenNotPaused(PAUSED_EXIT_WITHDRAWAL_QUEUE) {
+        bytes32 withdrawalRoot = calculateWithdrawalRoot(withdrawal);
+
+        require(
+            pendingWithdrawals[withdrawalRoot], 
+            "DelegationManager.completeQueuedAction: action is not in queue"
+        );
+
+        require(
+            slasher.canWithdraw(withdrawal.delegatedTo, withdrawal.startBlock, middlewareTimesIndex),
+            "DelegationManager.completeQueuedAction: pending action is still slashable"
+        );
+
+        require(
+            withdrawal.startBlock + withdrawalDelayBlocks <= block.number, 
+            "DelegationManager.completeQueuedAction: withdrawalDelayBlocks period has not yet passed"
+        );
+
+        require(
+            msg.sender == withdrawal.withdrawer, 
+            "DelegationManager.completeQueuedAction: only withdrawer can complete action"
+        );
+
+        if (receiveAsTokens) {
```

Comment:
would be good to not that `recieveAsTokens` is at the level of the withdrawal

Reply:
I added some documentation, please improve more if you think it needs it